main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -1156,6 +1079,175 @@ object GenExpression {
       // Emit code for the invocation of entails.
       visitor.visitMethodInsn(INVOKESTATIC, JvmName.Runtime.Fixpoint.Solver.toInternalName, "entails", "(Lflix/runtime/fixpoint/ConstraintSystem;Lflix/runtime/fixpoint/ConstraintSystem;)Z", false);
 
+    case Expression.FixpointFold(pred, init, f, constraints, tpe, loc) =>
+      // Add source line numbers for debugging.
+      addSourceLine(visitor, loc)
+
+      // push the initial loop index
+      visitor.visitInsn(ICONST_0)
+      // stack: [index]
+
+      // Get the initial accumulator value
+      readVar(init.sym, init.tpe, visitor)
+      // stack: [index, acc]
+
+      // Label for the fold loop
+      val loop = new Label
+      // Label for the exit of the loop
+      val exitLabel = new Label
+
+      // Begin of the loop
+      visitor.visitLabel(loop)
+      // stack: [acc, index]
+
+      // Rearrange the stack
+      visitor.visitInsn(DUP_X1) // stack: [index, acc, index]
+
+      // Get the projected constraint system
+      // We assume the constraint system has been projected beforehand (this is done in Simplifier)
+      readVar(constraints.sym, constraints.tpe, visitor)
+      // stack: [index, acc, index, constraints]
+      visitor.visitInsn(SWAP)
+      // stack: [index, acc, constraints, index]
+
+      // get the fact at the current index
+      visitor.visitMethodInsn(INVOKEVIRTUAL, JvmName.Runtime.Fixpoint.ConstraintSystem.toInternalName, "getFact",
+        "(I)Lflix/runtime/fixpoint/Constraint;", false)
+      // stack: [index, acc, fact]
+
+      visitor.visitInsn(DUP)
+      // stack: [index, acc, fact, fact]
+
+      // if fact is null, there is no more fact to handle, exit the loop
+      visitor.visitJumpInsn(IFNULL, exitLabel)
+      // stack: [index, acc, fact]
+
+      // Start constructing the tuple
+      // Extract the type of elements that are in the tuple
+      val tupleElmsTypes = root.relations(pred.sym.asInstanceOf[Symbol.RelSym] /* TODO: not clean */).attr.map(_.tpe)
```

Comment:
As part of this refactoring I am talking about, the root.relations map will disappear. 

Can this be accomplished without that information?

Reply:
The type of the elements in the relation is needed, but if needed it can be found in an earlier phase, except if the relations map completely disappears from all phases I guess?

Reply:
It is going to disappear. Can you get the arity from the type of some appropriate expression?

Reply:
The type of the tuple elements is actually just the type of the first argument to the folded function `f`, so it can be recovered that way here. That's fine for code generation if it has been previously type-checked. The problem then lies in the typer when checking that `f` is properly typed (Typer.scala, line 1251): `Type.mkTuple(program.relations(rel).attr.map(a => a.tpe)`

Reply:
Can I ask you the make the refactoring here? Then I will deal with the Typer later. (Not yet sure whta to do)

Reply:
Sure, doing it right now